[WIP] boards/stm32f2/f4/f7: rework clock configuration and expose clock configuration in kconfig

aabadie opened the pull request:
<!--
The RIOT community cares a lot about code quality.
Therefore, before describing what your contribution is about, we would like
you to make sure that your modifications are compliant with the RIOT
coding conventions, see https://github.com/RIOT-OS/RIOT/wiki/Coding-conventions.
-->

### Contribution description

Similar to #14921, #14923 and #14945, this PR is reworking the clock configuration and initialization of STM32f2/F4/F7 families:
- 3 clock sources can be used: HSE (if provided by the board), HSI, and PLL
- If selected, the PLL input clock source is HSE if the board provides one, HSI otherwise
- PLL prescalers are configurable
- MCO1 and MCO2 sources and prescalers are also fully (and easily) configurable
- APB1/2 bus clock dividers can also be configured easily
- the selection of the alternative 48MHz clock which, depending on the CPU family, can be retrieved from PLLI2S, PLLSAI or PLLQ, is now linked to several configuration parameters but also to the use of `MODULE_PERIPH_USBDEV`. This is the part of the PR I'm not 100% satisfied. Maybe @bergzand could have an opinion on that.

Compared to other similar PRs (for l0, l1, l4, f0, f1), it was more difficult to model the clock configuration in a generic fashion. So I had to anticipate a little bit on what things will look like in the end. Indeed, one can have several maximum clock speeds in the same family (84MHz, 100MHz, 168MHz, 180MHz for F4) and several source of clocks (PLLI2S and/or PLLSAI or PLL_Q) for USB. For the moment, in RIOT, PLL_Q, PLLI2S and PLLSAI are used to produce a 48MHz clock only for USB but that may change one day.

The current design is taken by this PR to model in a generic way is also important:
- clock configuration headers are called `cfg_clock_default_<clock speed>.h` (for example `cfg_clock_default_100.h`. This file only configures the PLL parameters used to reach the _clock speed_ used in the file name
- each clock configuration headers is divided in 3 parts:
  1. include a common header file which contain the common parameters for all families (clock sources, HSE and LSE default availability, PLLSAI/I2S parameters, etc). This header is called `cfg_clock_common.h`
  2. define the default PLL parameters used to compute SYSCLK. The default parameters can contain values for different HSE clock speed (default 8MHz, but sometimes 12MHz, 16MHz or 25MHz exist) and HSI (16MHz). This part also define the default APB1/2 prescalers (no bitfield, just the divider coef)
  3. include another header that contains the preprocessor logic to compute the clocks, CLOCK_CORECLOCK (which depends on the clock source) and CLOCK_APB1/2. In the long term, I'm planning to move that part to `cpu/stm32/stmclk.h` but I need other clock refactoring PRs to be in first. This header is called `cfg_clock_values.h` and must be included after all PLL parameters bus prescalers are known.
- each board must specify if it provides an HSE and/or LSE. If HSE is different from 8MHz, it must also be specified at that level.

All affected boards are updated and the parameters should remain the same.

The Kconfig part is not done yet but I'll add it later (that's why the PR is marked as WIP) or could be provided in a follow-up (in this case the WIP label can be removed

<!--
Put here the description of your contribution:
- describe which part(s) of RIOT is (are) involved
- if it's a bug fix, describe the bug that it solves and how it is solved
- you can also give more information to reviewers about how to test your changes
-->


### Testing procedure

- A green Murdock
- Functional tests of the affected boards (nucleo-f207zg, nucleo-f4*, nucleo-f7*, stm32f4*, stm32f7*, at least)
- Test the menu entry in make menuconfig:
```
make BOARD=<board name> -C examples/hello-world menuconfig
```
- Test the different clock modes using the following patch (required to print the clock values) and command lines:

<details><summary>patch</summary>

```diff
diff --git a/examples/hello-world/main.c b/examples/hello-world/main.c
index f51bf8c0a0..ee0e4ccbc5 100644
--- a/examples/hello-world/main.c
+++ b/examples/hello-world/main.c
@@ -20,6 +20,9 @@
  */
 
 #include <stdio.h>
+#include <inttypes.h>
+
+#include "board.h"
 
 int main(void)
 {
@@ -28,5 +31,12 @@ int main(void)
     printf("You are running RIOT on a(n) %s board.\n", RIOT_BOARD);
     printf("This board features a(n) %s MCU.\n", RIOT_MCU);
 
+    printf("Clock coreclock: %" PRIu32 "\n", (uint32_t)CLOCK_CORECLOCK);
+    printf("AHB clock: %" PRIu32 "\n", (uint32_t)CLOCK_AHB);
+    printf("APB1 clock: %" PRIu32 "\n", (uint32_t)CLOCK_APB1);
+#ifdef CLOCK_APB2
+    printf("APB2 clock: %" PRIu32 "\n", (uint32_t)CLOCK_APB2);
+#endif
+
     return 0;
 }
```

</details>

<details><summary>command lines</summary>

- use HSE:
```
CFLAGS=-DCONFIG_USE_CLOCK_HSE=1 make BOARD=<board name> -C examples/hello-world flash term
```
(this one will fail for boards that don't have an HSE)

- use HSI:
```
CFLAGS=-DCONFIG_USE_CLOCK_HSI=1 make BOARD=<board name> -C examples/hello-world flash term
```

- use PLL (default):
```
CFLAGS=-DCONFIG_USE_CLOCK_PLL=1 make BOARD=<board name> -C examples/hello-world flash term
```

</details>


<!--
Details steps to test your contribution:
- which test/example to compile for which board and is there a 'test' command
- how to know that it was not working/available in master
- the expected success test output
-->


### Issues/PRs references

Depends on #14923 because of split the of f0 and f1 out of `stmclk_fx.c`

<!--
Examples: Fixes #1234. See also #5678. Depends on PR #9876.

Please use keywords (e.g., fixes, resolve) with the links to the issues you
resolved, this way they will be automatically closed when your pull request
is merged. See https://help.github.com/articles/closing-issues-using-keywords/.
-->

vincent-d commented on cpu/stm32/stmclk/stmclk_f2f4f7.c:
```diff
@@ -0,0 +1,534 @@
+#endif
+
+/* Default is not configure MCO2 */
+#ifndef CONFIG_CLOCK_ENABLE_MCO1
```
`MCO2`

aabadie replied:
Fixed!

aabadie commented:
@vincent-d I have a problem with the MCO feature on F4: it's not working on stm32f429i-disc1 (I cannot test on other boards for the moment). Were you able to make it work on a nucleo board ?
In this PR, the MCO configuration is extended to be able to easily choose between the different clock sources (except LSE) and the MCO1 and MCO2 pins are also automatically initialized (I checked, they are always PA8 and PC9 on F2/F4/F7). For the MCO pins, do you think the autoinitialization makes sense ?

I'm asking you because you introduced the changes related to MCO in #7477.

fjmolinas commented:
@aabadie can you rebase, I haven't started reviewing and having it squashed will easy the review process.

aabadie commented:
As a side note, I took the nucleo-f207zg with me to test the MCO stuff. Will plug it back to tribe-ci when I have something that works (rather soonish :) ).

aabadie commented:
I found the issue with MCO. The MCO pins must be configured outside after the call to `_gpio_init_ain` in `cpu_init`:
https://github.com/RIOT-OS/RIOT/blob/2cfb4f49cbf9fa1afa222997b2f35590d1b30819/cpu/stm32/cpu_init.c#L157-L164

If MCO pins are initialized in `stmclk_clock_init`, they are reset right after so it's not possible to get the clock output.
@fjmolinas, do you think it would make sense to introduce a `mco_init` function that is called only if CONFIG_CLOCK_ENABLE_MCO1 or CONFIG_CLOCK_ENABLE_MCO2 is active ?

fjmolinas commented:
> If MCO pins are initialized in stmclk_init_sysclk, they are reset right after so it's not possible to get the clock output.

Can it be inverted otherwise? I think that should work as well no?

aabadie commented:
> Can it be inverted otherwise? I think that should work as well no?

Yes probably. I can try. Note that this change will affect a lot of hardware :)

aabadie commented:
> I can try. Note that this change will affect a lot of hardware :)

Inverting doesn't seem to introduce any issue. But MCO is now functional on F2 and F4.

aabadie commented:
@fjmolinas, do you think this change might deserve its own PR ?

fjmolinas commented:
> 
> 
>     I can try. Note that this change will affect a lot of hardware :)
> 
> Inverting doesn't seem to introduce any issue. But MCO is now functional on F2 and F4.

Lets go with that I'll take a look at consumption impact next week.

aabadie commented:
> can you rebase, I haven't started reviewing and having it squashed will easy the review process

This PR is based on #14923 so I rebased and squashed both (#14923 is already reviewed code-wise).

fjmolinas commented:
Can you rebase @aabadie?

fjmolinas commented on cpu/stm32/stmclk/stmclk_f2f4f7.c:
```diff
@@ -22,108 +22,444 @@
+#endif
+
+/* Check whether PLL must be enabled:
+  - When PLL is used system clock
```
```suggestion
  - When PLL is used as system clock
```

aabadie replied:
Fixed and directly squashed!

fjmolinas commented:
I ran some basic timing testing (`"tests/periph_timer tests/periph_rtc tests/periph_rtt tests/xtimer*"`) on one of each CPU_FAM, all good, results here:

https://ci.inria.fr/ci-riot-tribe/job/build-pipeline.jk/lastSuccessfulBuild/artifact/

fjmolinas reviewed (commented):
I'm testing that USB is still functional by running `tests/usbus_cdc_acm_stdio`, all seems OK!

<details> <summary>nucleo-f202zg</summary>

```
2020-09-29 14:56:04,964 #  text 123
2020-09-29 14:56:04,966 # 01234567890123456789012345678901234567890123456789012345678901234567890123456789
2020-09-29 14:56:04,967 # 0123456789012345678901234567890123456789012
version
2020-09-29 14:56:08,997 #  version
2020-09-29 14:56:08,998 # 2020.10-devel-1524-gb8b4a-wip/openwsn_radio_hal_csp
> reboot
2020-09-29 14:56:56,786 # Serial port disconnected, waiting to get reconnected...
2020-09-29 14:56:57,789 # Try to reconnect to /dev/ttyACM1 again...
2020-09-29 14:56:57,790 # Reconnected to serial port /dev/ttyACM1
help
2020-09-29 14:57:01,775 #  reboohelp
2020-09-29 14:57:01,776 # Command              Description
2020-09-29 14:57:01,776 # ---------------------------------------
2020-09-29 14:57:01,777 # text                 Generates long text for testing stdio buffer
2020-09-29 14:57:01,778 # reboot               Reboot the node
2020-09-29 14:57:01,779 # version              Prints current RIOT_VERSION
2020-09-29 14:57:01,780 # pm                   interact with layered PM subsystem
2020-09-29 14:57:01,781 # ps                   Prints information about running threads.
```
```
[21871.006715] usb 4-1.5.7: New USB device found, idVendor=1209, idProduct=7d00, bcdDevice= 0.00
[21871.006718] usb 4-1.5.7: New USB device strings: Mfr=3, Product=2, SerialNumber=0
[21871.006720] usb 4-1.5.7: Product: USB device
[21871.006721] usb 4-1.5.7: Manufacturer: RIOT-os.org
[21871.015428] cdc_acm 4-1.5.7:1.0: ttyACM1: USB ACM device
[21979.716742] xhci_hcd 0000:00:14.0: WARN Cannot submit Set TR Deq Ptr
[21979.716744] xhci_hcd 0000:00:14.0: A Set TR Deq Ptr command is pending.
[21979.948452] usb 4-1.5.7: USB disconnect, device number 34
[21979.950857] cdc_acm 4-1.5.7:1.0: failed to set dtr/rts
[21980.208127] usb 4-1.5.7: new full-speed USB device number 35 using xhci_hcd
[21980.330260] usb 4-1.5.7: New USB device found, idVendor=1209, idProduct=7d00, bcdDevice= 0.00
[21980.330264] usb 4-1.5.7: New USB device strings: Mfr=3, Product=2, SerialNumber=0
[21980.330267] usb 4-1.5.7: Product: USB device
[21980.330269] usb 4-1.5.7: Manufacturer: RIOT-os.org
[21980.338732] cdc_acm 4-1.5.7:1.0: ttyACM1: USB ACM device
```
</details>


<details> <summary>nucleo-f746zg</summary>

```
2020-09-29 14:58:35,452 # Reconnected to serial port /dev/ttyACM1
2020-09-29 14:58:35,453 #  main(): This is RIOT! (Version: 2020.10-devel-1524-gb8b4a-wip/openwsn_radio_hal_csp)
2020-09-29 14:58:35,453 # RIOT USB CDC ACM shell test
> text 10
2020-09-29 14:58:50,997 #  text 10
2020-09-29 14:58:50,997 # 0123456789
> reboot
2020-09-29 14:58:52,505 # Serial port disconnected, waiting to get reconnected...
2020-09-29 14:58:53,509 # Try to reconnect to /dev/ttyACM1 again...
2020-09-29 14:58:53,510 # Reconnected to serial port /dev/ttyACM1
2020-09-29 14:58:53,511 #  reboomain(): This is RIOT! (Version: 2020.10-devel-1524-gb8b4a-wip/openwsn_radio_hal_csp)
2020-09-29 14:58:53,512 # RIOT USB CDC ACM shell test
>

```
```
[22076.460253] usb 4-1.5.7: USB disconnect, device number 37
[22077.727155] usb 4-1.5.7: new full-speed USB device number 38 using xhci_hcd
[22077.852551] usb 4-1.5.7: New USB device found, idVendor=1209, idProduct=7d00, bcdDevice= 0.00
[22077.852552] usb 4-1.5.7: New USB device strings: Mfr=3, Product=2, SerialNumber=0
[22077.852553] usb 4-1.5.7: Product: USB device
[22077.852554] usb 4-1.5.7: Manufacturer: RIOT-os.org
[22077.860336] cdc_acm 4-1.5.7:1.0: ttyACM1: USB ACM device
[22095.439541] xhci_hcd 0000:00:14.0: WARN Cannot submit Set TR Deq Ptr
[22095.439543] xhci_hcd 0000:00:14.0: A Set TR Deq Ptr command is pending.
[22095.441757] xhci_hcd 0000:00:14.0: WARN Cannot submit Set TR Deq Ptr
[22095.441758] xhci_hcd 0000:00:14.0: A Set TR Deq Ptr command is pending.
[22095.660573] usb 4-1.5.7: USB disconnect, device number 38
[22095.664372] cdc_acm 4-1.5.7:1.0: failed to set dtr/rts
[22095.910550] usb 4-1.5.7: new full-speed USB device number 39 using xhci_hcd
[22096.032109] usb 4-1.5.7: New USB device found, idVendor=1209, idProduct=7d00, bcdDevice= 0.00
[22096.032115] usb 4-1.5.7: New USB device strings: Mfr=3, Product=2, SerialNumber=0
[22096.032119] usb 4-1.5.7: Product: USB device
[22096.032122] usb 4-1.5.7: Manufacturer: RIOT-os.org
[22096.043939] cdc_acm 4-1.5.7:1.0: ttyACM1: USB ACM device

```
</details>

<details> <summary>nucleo-f429zi </summary>

```
[22755.628613] usb 4-1.5.3: USB disconnect, device number 43
[22755.898024] usb 4-1.5.3: new full-speed USB device number 44 using xhci_hcd
[22756.020393] usb 4-1.5.3: New USB device found, idVendor=1209, idProduct=7d00, bcdDevice= 0.00
[22756.020400] usb 4-1.5.3: New USB device strings: Mfr=3, Product=2, SerialNumber=0
[22756.020404] usb 4-1.5.3: Product: USB device
[22756.020406] usb 4-1.5.3: Manufacturer: RIOT-os.org
[22756.028794] cdc_acm 4-1.5.3:1.0: ttyACM0: USB ACM device
[22773.804585] usb 4-1.5.3: USB disconnect, device number 44
[22774.053076] usb 4-1.5.3: new full-speed USB device number 45 using xhci_hcd
[22774.175080] usb 4-1.5.3: New USB device found, idVendor=1209, idProduct=7d00, bcdDevice= 0.00
[22774.175086] usb 4-1.5.3: New USB device strings: Mfr=3, Product=2, SerialNumber=0
[22774.175090] usb 4-1.5.3: Product: USB device
[22774.175092] usb 4-1.5.3: Manufacturer: RIOT-os.org
[22774.184744] cdc_acm 4-1.5.3:1.0: ttyACM0: USB ACM device
[22790.188695] usb 4-1.5.3: USB disconnect, device number 45
[22790.191039] cdc_acm 4-1.5.3:1.0: failed to set dtr/rts
[22790.436260] usb 4-1.5.3: new full-speed USB device number 46 using xhci_hcd
[22790.566561] usb 4-1.5.3: New USB device found, idVendor=1209, idProduct=7d00, bcdDevice= 0.00
[22790.566566] usb 4-1.5.3: New USB device strings: Mfr=3, Product=2, SerialNumber=0
[22790.566569] usb 4-1.5.3: Product: USB device
[22790.566571] usb 4-1.5.3: Manufacturer: RIOT-os.org
[22790.576001] cdc_acm 4-1.5.3:1.0: ttyACM0: USB ACM device

```
```
2020-09-29 15:10:21,854 # main(): This is RIOT! (Version: 2020.10-devel-1619-g54d87-pr-14946)
2020-09-29 15:10:21,855 # RIOT USB CDC ACM shell test
> help
2020-09-29 15:10:24,221 #  help
2020-09-29 15:10:24,222 # Command              Description
2020-09-29 15:10:24,223 # ---------------------------------------
2020-09-29 15:10:24,224 # text                 Generates long text for testing stdio buffer
2020-09-29 15:10:24,225 # reboot               Reboot the node
2020-09-29 15:10:24,225 # version              Prints current RIOT_VERSION
2020-09-29 15:10:24,226 # pm                   interact with layered PM subsystem
2020-09-29 15:10:24,227 # ps                   Prints information about running threads.
> reboot
2020-09-29 15:10:27,066 # Serial port disconnected, waiting to get reconnected...
2020-09-29 15:10:28,069 # Try to reconnect to /dev/ttyACM0 again...
2020-09-29 15:10:28,071 # Reconnected to serial port /dev/ttyACM0
help
2020-09-29 15:10:44,383 #  reb: This is RDhelp
2020-09-29 15:10:44,385 # shell: command not found: main():

```
</details>

I also force SAI to be enabled by running  `CFLAGS=-DCONFIG_CLOCK_PLL_N=120 BOARD=nucleo-f746zg make -C tests/usbus_cdc_acm_stdio/ flash -j` and the usb device still showed up.

I also changed the same config for `nucleo-f429zi` and since the only possible source for the 48MHZ clock is the PLL , I get an error.

fjmolinas commented:
Note there is on checking of the range values for `PLL_%`, but this is documented in the tracking issue.

fjmolinas commented:
I also tested the AIN change did not impact power consumption on `nucleo-l152re`, did not test on other `BOARD`s.

fjmolinas commented on boards/common/stm32/include/f2f4f7/cfg_clock_default_168.h:
```diff
@@ -0,0 +1,90 @@
+ * frequency of 180MHz.
+ * @{
+ */
+/* The following parameters configure a 168MHz system clock with HSE
+   (8MHz, 12MHz or 16MHz) or HSI (16MHz) as PLL input clock */
```
Why 168Mhz?

aabadie replied:
It's because, in RIOT, we use 168MHz instead of 180MHz for compatibility with USB 48MHz required clock: it's not possible to configure a 48MHz on PLLQ when PLLP is configured for 180MHz (no common divisor).
There's the same issue with M4 that can have a 100MHz max core clock, in RIOT, they are configured to 96MHz.

See #15043 which improves the situation: use 100/180MHz by default, except when `periph_usbdev` is used. This PR also removes the useless `cfg_clock_default_96.h` and `cfg_clock_default_168.h` default clock configuration headers.

fjmolinas replied:
ok

aabadie commented on cpu/stm32/stmclk/stmclk_f2f4f7.c:
```diff
@@ -22,108 +22,444 @@
+/* Compute the bitfields for the PLL configuration */
+#define PLL_M                       (CONFIG_CLOCK_PLL_M << RCC_PLLCFGR_PLLM_Pos)
+#define PLL_N                       (CONFIG_CLOCK_PLL_N << RCC_PLLCFGR_PLLN_Pos)
+#define PLL_P                       (((CONFIG_CLOCK_PLL_P / 2) - 1) << RCC_PLLCFGR_PLLP_Pos)
```
[It was already like this](https://github.com/RIOT-OS/RIOT/pull/14946/files#diff-ad72ec4aaf9e02da8272f42b692417f6L103). It's just code that was move around.

fjmolinas replied:
ok...

fjmolinas reviewed (approved):
ACK
